### PR TITLE
Fix strobes from overflooding *conn documents when strobing is cumulative

### PR DESCRIPTION
**Pull request by caffeinatedpixel**
Resolves an issue where uconn, uconnProxy, and SNIconn documents would grow over the maximum Mongo BSON size (16MB) if the connection became a strobe due to its cumulative connection count over all chunks.

- Adds a "siphon" step to the analysis process for beacons, beaconSNI, and beaconProxy which runs after the dissector and makes necessary updates prior to sorting & analysis
  -  I chose to add a step to each package as opposed to creating a standalone strobe package due to the need to have awareness of the cumulative connection count/strobe status based off of the cumulative connection count. This avoids having to look up each parsed uconn, SNIconn, and uconnProxy again.
 The siphon concept comes from a coffee siphon maker, where some liquid is evaporated in the top chamber and the rest is drained into the bottom chamber, resulting in coffee. The behavior here is similar, where *conns are dissected, some are evaporated (strobes) and some are drained into the rest of the pipeline (non-strobes). This allows for the analysis process to not care about strobes as the analysis channel will never receive them. 

The process is shown below:

```mermaid
flowchart TD
	A[Uconn] --> B[Analyzer] --> C(This CID's conn count is over the strobe limit?)
	C --> |Yes| D[Set as strobe and set timestamps to empty]
	C --> |No| E[Set strobe to false and push timestamps to dat]
	F[Beacons] --> G[Dissector]
	D --> F
	E --> F
        G --> X[Get uconn if it is not a strobe according to mongo]
	X --> H(The sum of connections over all CIDs is over the strobe limit?)
	H --> |Yes| M[Siphon Evaporate \n - Remove this chunk's timestamps and bytes \n - Set strobe status\n - Remove uconn from beacon table]
	H --> |No| N[Siphon Drain - Pass uconn through to sorter]
	N --> I[Sorter]
	I --> P[Analyzer]
	P --> O[Writer]
```
So now, whenever a connection is a strobe based on its cumulative connection count, its timestamps and bytes that were added to the `dat[chunk]` document are removed. If a connection is a strobe based on the current chunk's connection count, then uconn never writes out the timestamp and byte arrays to begin with.

Tested the time taken for analysis over 5 large chunks on this PR and master. The elapsed time only differed by a fraction of a second.
Tested ingestion on a rolling dataset where the first chunk didn't meet strobe criteria, and ensured the subsequent chunks that did meet it did not contain any timestamp or bytes arrays.

### CODE REVIEW

**Review by Zalgo2462 — COMMENTED** [review]
I used the notes in https://github.com/activecm/AC-Hunter/pull/2184#pullrequestreview-1211935024 to create a test dataset which triggers the testing conditions (big thank you btw). 

Everything ran well :tada: 

---

I think there is a bit of a weird behavior when a strobe goes back to being a beacon. If the `dat` subdocuments with the very large connection counts roll out of the dataset, the subsequent `dat` subdocuments might not have their `ts` and `byte` arrays. Since the large `count` fields no longer exist, the siphon will then send the unique connection down the drain to beacon analysis. So, down the line, we could end up with a beacon where the number of `ts` and `byte` entries does not match the total connection count.  This would correct itself over time though.

However, I believe this is the intended behavior?

---

Would it make sense to just remove all of the `bytes` and `ts` fields instead of just the latest chunk's in the evaporator?

---

I would prefer to queue the dataset operations up in the bulk writer instead of running them one-off inside the evaporator. But there might be complications with that (see the other comment). Lmk what you think. If it's a PITA, I'll approve the current code.

File: pkg/beacon/siphon.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		defer ssn.Close()
+
+		for data := range s.siphonChannel {
+			// if there are evaporation tasks to complete, run through all of them with normal (non-bulk) queries
+			// because mgo does not support updating with array filters in bulk operations
+			if data.Evaporate != nil {
+				for _, action := range data.Evaporate {
+					// run UpdateWithArrayFilters if arrayFilters are set
+					if action.arrayFilters != nil {
+						info, err := ssn.DB(s.db.GetSelectedDB()).C(action.collection).UpdateWithArrayFilters(
```
[the remark this answers is not in the document]
Comment (caffeinatedpixel) [reply pkg/beacon/siphon.go:81]: I tested imports against updating all chunks vs just the current chunk, and the results appear to be identical, so I don't foresee any issues with making the array removals for every chunk. This removes the need for the positional operator query you provided or the need for arrayFilters.

For SNIconns, I have run into an issue trying to remove the arrays for `tls` and `http` using the bulk writer. I am not sure if the bulk writer cannot handle updating nested subarrays or what the deal is, but I don't have the bandwidth at the moment to muck around with it. I was able to update `uconn` and `uconnProxy` using the bulk writer with `{"$unset": {"dat.$[].ts": "" ...}`, but this does not work for `SNIconn` and nor does a few dozen different variations and key selectors. 

I can push up the changes for `uconn` and `uconnproxy` to use the bulk writer & remove the arrays for all cids. Unless there's another way to deal with updating `SNIconn`, I'd vote to leave it as is.

[the remark this answers is not in the document]
Comment (Zalgo2462) [reply pkg/beacon/siphon.go:81]: Please push up the changes for uconn and uconnproxy. I will look into changing the SNIconn tonight. If I can't come up with anything tonight, I agree we should leave it as is.

[the remark this answers is not in the document]
Comment (Zalgo2462) [reply pkg/beacon/siphon.go:81]: As I understand it, `$` only matches a single document, and `$[]` will match all of the dat documents if an array filter isn't supplied. I think we can remove the `http` and `tls` records in `SNIconn` if we split the update up into two updates, where each one removes a single subdocument. I'll make these changes tonight.

**Comment by Zalgo2462** [comment]
I've pushed up my changes.

The first commit addresses removing the `http` and `tls` timestamps and bytes from the `SNIconn` subdocuments using the bulk writer. 

The second and third commits change the siphons in IP beacons and Proxy beacons packages back to only removing the `ts` and `bytes` fields for the current chunk's subdocument rather than removing them from all of the past chunks' subdocuments. 

I've tested these changes using the almost-strobes dataset and importing the first three hours. For the SNI and IP beacons, only the first chunk's subdocuments contain `ts` and `bytes` fields. For the Proxy beacons, the first two chunk's subdocuments contain these fields. This matches my expectations based on where my strobe limit is set (86400). 

Please test these changes on your end and let me know if they look good to you. If so, lmk and I'll get this pulled in.

**Comment by caffeinatedpixel** [comment]
The code looks good.
The change in the beaconsni siphon that changes the setting of the `dat.merged` entry as an object instead of an array does fix the lookup of the `merged.strobe` property in the beaconsni dissector (which appears to be broken in master), but it causes an issue where the beaconsni dissector doesn't retrieve the sniconn entry when
- conn is a merged strobe in a previous chunk 
- tls is a strobe and http isn't (and vice versa)
- a previous chunk has tls or/and http as a strobe (is an issue if current chunk is not a strobe for either tls or http)

I pushed fixes for this issue here: https://github.com/activecm/rita/tree/2157-bulk-fixes-sni 
It changes the query for sniconns in the beaconsni dissector so that only the current cid is matched for strobeness

Testing this fix along with the changes in #770 worked well and resulted in the proper data for all 24 chunks with very small document sizes.

**Comment by Zalgo2462** [comment]
The changes in https://github.com/activecm/rita/commit/d9845b07197d34cb26e99c35c44054c8e83c96bb LGTM. 

How do you want to pull off the merge. Do you want to apply that commit here and then separately review #770? Or, do you want to pull 770 in here and then add that commit on top?

**Comment by caffeinatedpixel** [comment]
I'd prefer to review #770 as a PR into this PR and then apply https://github.com/activecm/rita/commit/d9845b07197d34cb26e99c35c44054c8e83c96bb and test all three branches at once/together

**Comment by Zalgo2462** [comment]
I've merged #770 and https://github.com/activecm/rita/commit/d9845b07197d34cb26e99c35c44054c8e83c96bb into this branch. 

I will give it one last run through for testing now that the merging is finished. If that looks good, I'll go ahead and pull it all in.

**Review by Zalgo2462 — CHANGES_REQUESTED** [review]
I think there is an error in the https://github.com/activecm/rita/commit/d9845b07197d34cb26e99c35c44054c8e83c96bb patch. Even when both `http.strobe` and `tls.strobe` are set to `true`, the document is processed by the SNIConn module and a `merged.strobe` document is getting inserted. 

Would it make sense to replace
```go
matchNoStrobeKey["dat"] = bson.M{
	"$elemMatch": bson.M{
		"cid": d.chunk,
		"$or": []interface{}{
			bson.M{"tls.strobe": bson.M{"$ne": true}},
			bson.M{"http.strobe": bson.M{"$ne": true}},
			// merged must be added to this filter if it is ever set prior to
			// this pipeline in the future
			// "merged.strobe": bson.M{"$ne": true},
		},
	}}
```

with 

```go
matchNoStrobeKey["dat"] = bson.M{
	"$elemMatch": bson.M{
		"cid": d.chunk,
		"$or": []interface{}{
			bson.M{"tls.strobe": false},
			bson.M{"http.strobe": false},
			// merged must be added to this filter if it is ever set prior to
			// this pipeline in the future
			// bson.M{"merged.strobe": false},
		},
	}}
```

I think the `{"$ne": true}` bit is matching the non-existent fields in the other subdocuments.

This patch seemed to work to fix the issue for me.

File: database/writer.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+// close waits for the write threads to finish
+func (w *MgoBulkWriter) Close() {
+	close(w.writeChannel)
+	w.writeWg.Wait()
+}
+
+// start kicks off a new write thread
+func (w *MgoBulkWriter) Start() {
```
Comment (caffeinatedpixel) [inline database/writer.go:107]: Please add comments throughout this function for clarity & readability

File: database/writer.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	return &MgoBulkWriter{
+		db:           db,
+		conf:         conf,
+		log:          log,
+		writeChannel: make(chan BulkChanges),
+		writeWg:      new(sync.WaitGroup),
+		writerName:   writerName,
+		unordered:    unorderedWritesOK,
+		maxBulkCount: 500,
+		maxBulkSize:  15 * 1000 * 1000,
```
[the remark this answers is not in the document]
Comment (Zalgo2462) [reply database/writer.go]: Added

File: database/writer.go (hunk 2 above)
Comment (Zalgo2462) [reply database/writer.go:109]: Added

**Comment by caffeinatedpixel** [comment]
Code looks good. Tested with generated data. I'm good with your changes from #770.

**Review by Zalgo2462 — APPROVED** [review]
Everything looks good to me. :heavy_check_mark:

**Pull request merged**